### ISSUE TITLE
RyuJIT: Mishandling of subrange assertion for rewritten call parameter

### ISSUE THREAD

**Issue by jakobbotsch** [dotnet/coreclr#19558]
The example is:
```csharp
// Debug: Outputs 0
// Release: Outputs 65536
public class Program
{
    static short s_19;
    public static void Main()
    {
        M75(-1);
    }

    static void M75(short arg0)
    {
        try
        {
            arg0 = -1;
        }
        finally
        {
            arg0 &= 1;
        }

        arg0 = s_19;
        short var11 = arg0;
        System.Console.WriteLine(0 - var11);
    }
}
```
(remember to set `COMPlus_TieredCompilation` to 0 in release).

Similarly to #18867, RyuJIT reloads the argument as a 32-bit value without handling the upper bits correctly:
```asm
G_M59072_IG01:
       55                   push     rbp
       4883EC10             sub      rsp, 16
       488D6C2410           lea      rbp, [rsp+10H]
       488965F0             mov      qword ptr [rbp-10H], rsp
       894D10               mov      dword ptr [rbp+10H], ecx

G_M59072_IG02:
       480FBF0DE11FEEFF     movsx    rcx, word  ptr [reloc classVar[0x7d865280]]
       66894D10             mov      word  ptr [rbp+10H], cx
       8B4D10               mov      ecx, dword ptr [rbp+10H]
       F7D9                 neg      ecx
       48B82022987DFF7F0000 mov      rax, 0x7FFF7D982220

G_M59072_IG03:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       48FFE0               rex.jmp  rax
```

@mikedn has analyzed it here: https://github.com/dotnet/coreclr/issues/18867#issuecomment-413246494

**Comment by mikedn** [comment]
Hrm, apparently this does not reproduce anymore. However, the initial example in https://github.com/dotnet/coreclr/issues/18867#issuecomment-411234557 still fails, though with a different value - -32766 instead of 1.

**Comment by mikedn** [comment]
Yeah, so with or without/different repro the problem persists - `fgMorphLocalVar` uses subrange assertions to avoid inserting the "normalize on load" cast.

That can't work correctly because such subrange assertions can be generated by an earlier small int assignment, if you store 16 bits to a variable then it's obvious that you must insert the normalization cast to get 32 bits out of a subsequent use of the variable.

Unfortunately, fixing this results in a small CQ regression:
```
Total bytes of diff: 59 (0.00% of base)
    diff is a regression.
Top file regressions by size (bytes):
          21 : System.Text.RegularExpressions.dasm (0.02% of base)
          12 : System.Private.Xml.dasm (0.00% of base)
          11 : System.Private.CoreLib.dasm (0.00% of base)
           8 : Microsoft.CodeAnalysis.dasm (0.00% of base)
           3 : System.Private.Uri.dasm (0.00% of base)
8 total files with size differences (0 improved, 8 regressed), 121 unchanged.
Top method regressions by size (bytes):
          21 ( 4.82% of base) : System.Text.RegularExpressions.dasm - RegexCharClass:AddLowercaseRange(ushort,ushort,ref):this
          12 ( 2.80% of base) : System.Private.Xml.dasm - XmlSqlBinaryReader:ParseMB32_(ubyte):int:this
           8 ( 3.29% of base) : Microsoft.CodeAnalysis.dasm - DesktopAssemblyIdentityComparer:IsFrameworkAssembly(ref):bool
           4 (10.00% of base) : System.Private.CoreLib.dasm - Math:Abs(short):short
           4 (10.00% of base) : System.Private.CoreLib.dasm - Math:Abs(byte):byte
Top method regressions by size (percentage):
           4 (10.00% of base) : System.Private.CoreLib.dasm - Math:Abs(short):short
           4 (10.00% of base) : System.Private.CoreLib.dasm - Math:Abs(byte):byte
          21 ( 4.82% of base) : System.Text.RegularExpressions.dasm - RegexCharClass:AddLowercaseRange(ushort,ushort,ref):this
           8 ( 3.29% of base) : Microsoft.CodeAnalysis.dasm - DesktopAssemblyIdentityComparer:IsFrameworkAssembly(ref):bool
          12 ( 2.80% of base) : System.Private.Xml.dasm - XmlSqlBinaryReader:ParseMB32_(ubyte):int:this
10 total methods with size differences (0 improved, 10 regressed), 112828 unchanged.
```
That's because there are valid cases when the normalization cast can be eliminated - when the variable is enregistered all 32 bits are valid, not only the lower 8/16 bits.

So perhaps a better solution is to make "normalize on load" variables also "normalize on store", that will make the subrange assertion valid since we generate a 32 bit store. But there's at least one case where we can't actually do that - dependent promoted struct fields. Hrm.

**Comment by mikedn** [comment]
> Hrm, apparently this does not reproduce anymore.

The example in this issue has actually been fixed by #21184 

> However, the initial example in #18867 (comment) still fails, though with a different value - -32766 instead of 1.

And #21184 is also the reason why that example generates a different value now. Of course, it's still incorrect so there's still a bug to be fixed somewhere.

**Comment by mikedn** [comment]
@jakobbotsch Could you update the example in this issue to the one in https://github.com/dotnet/coreclr/issues/18867#issuecomment-411234557 ? 

As mentioned above, the current example was actually fixed by another change but the example in that comment still fails. And anyway that comment example is the "original" example that started this, the current one is supposed to be a simplification but it turns out that it is too simple :)

I'm still not 100% sure where the bug really is. It could be that it's incorrect for morph to elide load normalization in some cases but it also looks like my magic division code in lowering is broken in this case. I think I'll just fix the later for now.

**Comment by jakobbotsch** [comment]
@mikedn Done. Not sure if it helps, but there are probably some way simpler examples over in Fuzzlyn. For example:
https://github.com/jakobbotsch/Fuzzlyn/commit/f52cb42fe9f7eb6487a86da589250c4d7a7ba776
https://github.com/jakobbotsch/Fuzzlyn/commit/89eb6a39348756eb2dbe56bdaad2ac1e02daf1e9
https://github.com/jakobbotsch/Fuzzlyn/commit/dcedff2af5bf5dec5e1dfa1e60f607b8dbb56433
These have not been tested on Windows, however. But you can check out the [ubuntu-x64-examples](https://github.com/jakobbotsch/Fuzzlyn/commits/ubuntu-x64-examples) if you want to hunt for some simpler ones.

**Comment by mikedn** [comment]
Thanks! 

I tested all those 3 examples, they all fail even on Windows as well and they all pass after the fix. I ended up including the original example as test, due to its complexity it seems like it has a better chance to hit the bug. It's a rather elusive bug, at least because it requires spilled/not enregistered variables. That's why so far all examples that show the issue have exception handling in them.

I'm not sure that there aren't other bugs related to load normalization, it's a rather convoluted affair. And if there aren't anymore bugs then there may be CQ issues. While in the beginning I was thinking that morph is perhaps wrong not to insert the normalization cast in some cases, now it looks to me that in fact it's too aggressive at doing that.

But I'm 99.99% sure that the bug we're after now is actually in division lowering and that's what I fixed.
